spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/operations/SqsTemplate.java

```diff
@@ -0,0 +1,866 @@
+ * limitations under the License.
+ */
+package io.awspring.cloud.sqs.operations;
+
+import io.awspring.cloud.sqs.MessageHeaderUtils;
+import io.awspring.cloud.sqs.QueueAttributesResolver;
+import io.awspring.cloud.sqs.SqsAcknowledgementException;
+import io.awspring.cloud.sqs.listener.QueueAttributes;
+import io.awspring.cloud.sqs.listener.QueueNotFoundStrategy;
+import io.awspring.cloud.sqs.listener.SqsHeaders;
+import io.awspring.cloud.sqs.listener.acknowledgement.AcknowledgementCallback;
+import io.awspring.cloud.sqs.support.converter.MessageAttributeDataTypes;
+import io.awspring.cloud.sqs.support.converter.MessageConversionContext;
+import io.awspring.cloud.sqs.support.converter.MessagingMessageConverter;
+import io.awspring.cloud.sqs.support.converter.SqsMessageConversionContext;
+import io.awspring.cloud.sqs.support.converter.SqsMessagingMessageConverter;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.Assert;
+import software.amazon.awssdk.services.sqs.SqsAsyncClient;
+import software.amazon.awssdk.services.sqs.model.BatchResultErrorEntry;
+import software.amazon.awssdk.services.sqs.model.DeleteMessageBatchRequest;
+import software.amazon.awssdk.services.sqs.model.DeleteMessageBatchRequestEntry;
+import software.amazon.awssdk.services.sqs.model.DeleteMessageBatchResponse;
+import software.amazon.awssdk.services.sqs.model.DeleteMessageBatchResultEntry;
+import software.amazon.awssdk.services.sqs.model.Message;
+import software.amazon.awssdk.services.sqs.model.MessageAttributeValue;
+import software.amazon.awssdk.services.sqs.model.MessageSystemAttributeName;
+import software.amazon.awssdk.services.sqs.model.MessageSystemAttributeNameForSends;
+import software.amazon.awssdk.services.sqs.model.MessageSystemAttributeValue;
+import software.amazon.awssdk.services.sqs.model.QueueAttributeName;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageRequest;
+import software.amazon.awssdk.services.sqs.model.ReceiveMessageResponse;
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequest;
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchRequestEntry;
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchResponse;
+import software.amazon.awssdk.services.sqs.model.SendMessageBatchResultEntry;
+import software.amazon.awssdk.services.sqs.model.SendMessageRequest;
+
+public class SqsTemplate<T> extends AbstractMessagingTemplate<T, Message>
```

Comment:
I am not convinced with the generic in this class (and same for `SqsOperations` interface). It feels a bit awkward to use autoconfigured `SqsTemplate<Object>`. 

```java
private final SqsTemplate<Object> sqs;
...
SendResult<Object> hello = sqs.send("my-queue", "hello");
Message<Object> result = hello.message(); // returns Message<Object> even though we have passed string
```

Could we change in a way that class does not have generics but its methods still do? So that it works in following way:

```java
private final SqsTemplate sqs;
...
SendResult<String> result = sqs.send("my-queue", "hello");
Message<String> message = hello.message();

SendResult<Person> result2 = sqs.send("my-queue", new Person("john"));
Message<Person> message2 = hello.message();
```

Reply:
Great call again, thanks!

> I am not convinced with the generic in this class (and same for `SqsOperations` interface). It feels a bit awkward to use autoconfigured `SqsTemplate<Object>`.

Yeah, I had some doubts about generics throughout the project and removed a lot of what `Spring Kafka` has. Specially since we can't lookup the generics at runtime to e.g. figure out to which class to convert to.

But then I figured out that this generics is important for type safety at compile time. Let's say you have a queue that requires special configuration, and hence you have a different `Template` with maybe even a different `SqsAsyncClient`. You'd probably want to ensure at compile time that the correct `Template` is being used, and for that you can type this generics accordingly. Makes sense to you?

Good thing is I think we can do both the generics and your suggestion!

> ```java
> private final SqsTemplate sqs;
> ...
> SendResult<String> result = sqs.send("my-queue", "hello");
> Message<String> message = hello.message();
> 
> SendResult<Person> result2 = sqs.send("my-queue", new Person("john"));
> Message<Person> message2 = hello.message();
> ```

I'm thinking maybe we can do something similar as I did for the `receive` methods, and type it as `<P extends T>`. This way for the `<Object>` template we can use any class as you suggested, and for specifically-typed templates we can use any subclass.

How does that sound?

Reply:
Just to be clear, I meant the `payloadClass` parameter in the `SqsReceiveOptions` interface.

Reply:
> Let's say you have a queue that requires special configuration, and hence you have a different Template with maybe even a different SqsAsyncClient. You'd probably want to ensure at compile time that the correct Template is being used, and for that you can type this generics accordingly. Makes sense to you?

This would mean that a type determines a queue which may not be true. Same type can be sent to multiple queues. Using different `SqsTemplate` beans can be achieved with qualifiers - this does not give the compile time guarantees but let's face it many things in Spring don't give you compile time guarantees. 

> private final SqsTemplate sqs;

This gives a warning about raw use of parameterized type. I believe most people will just use `SqsTemplate<Object>`  and in such case I think it causes more headache than problems it solves. I am opting in for just `SqsTemplate` with parameterized methods without parameterized type.